File: sticky-notes-manager/src/core/note-store.js
// 便签管家 - 核心存储与逻辑层
// 纯 Node.js 模块，零依赖，可被 main.js 和 test.js 共用

'use strict';

const fs = require('fs');
const path = require('path');

// 默认便签数据文件路径（在用户数据目录下）
function defaultDataPath() {
  const home = process.env.APPDATA || process.env.HOME || process.env.USERPROFILE || '.';
  return path.join(home, 'sticky-notes-manager', 'notes.json');
}

// 生成唯一 ID（时间戳 + 随机串，避免碰撞）
function generateId() {
  return 'n_' + Date.now().toString(36) + '_' + Math.random().toString(36).slice(2, 8);
}

// 颜色标签枚举（苹果白风格配色）
const COLORS = {
  default: { name: '默认', hex: '#f5f5f7', dot: '#8e8e93' },
  blue:    { name: '蓝色', hex: '#e3f0ff', dot: '#007aff' },
  green:   { name: '绿色', hex: '#e8f8ec', dot: '#34c759' },
  yellow:  { name: '黄色', hex: '#fff9e0', dot: '#ffcc00' },
  orange:  { name: '橙色', hex: '#fff0e0', dot: '#ff9500' },
  pink:    { name: '粉色', hex: '#ffe8ef', dot: '#ff2d55' },
  purple:  { name: '紫色', hex: '#f3e8ff', dot: '#af52de' }
};

// 分类枚举
const CATEGORIES = ['工作', '个人', '灵感', '待办', '其他'];

// 创建便签对象（带默认值）
function createNote(data) {
  const now = Date.now();
  return {
    id: data.id || generateId(),
    title: (data.title || '').trim(),
    content: (data.content || '').trim(),
    color: data.color && COLORS[data.color] ? data.color : 'default',
    category: data.category && CATEGORIES.includes(data.category) ? data.category : '其他',
    pinned: !!data.pinned,
    createdAt: data.createdAt || now,
    // 保留导入/恢复数据中的原始 updatedAt；新建便签使用 now
    updatedAt: typeof data.updatedAt === 'number' && data.updatedAt > 0 ? data.updatedAt : now
  };
}

// 创建回收站便签对象（在普通便签基础上增加 deletedAt）
function createTrashNote(data) {
  const note = createNote(data);
  note.deletedAt = data.deletedAt || Date.now();
  return note;
}

// 读取全部便签
function loadNotes(dataPath) {
  return loadAll(dataPath).notes;
}

// 读取便签 + 回收站（统一入口，向后兼容 v1 数据）
function loadAll(dataPath) {
  const filePath = dataPath || defaultDataPath();
  try {
    if (!fs.existsSync(filePath)) {
      return { notes: [], trash: [] };
    }
    const raw = fs.readFileSync(filePath, 'utf-8');
    const data = JSON.parse(raw);
    const notes = Array.isArray(data.notes) ? data.notes.map(n => createNote(n)) : [];
    // v2 新增 trash 字段；v1 旧文件没有 trash，默认空数组
    const trash = Array.isArray(data.trash) ? data.trash.map(n => createTrashNote(n)) : [];
    return { notes, trash };
  } catch (e) {
    return { notes: [], trash: [] };
  }
}

// 保存全部便签
function saveNotes(notes, dataPath) {
  // 兼容旧调用：保留已有回收站数据
  const existing = loadAll(dataPath);
  return saveAll(notes, existing.trash, dataPath);
}

// 保存便签 + 回收站（v2 数据格式）
function saveAll(notes, trash, dataPath) {
  const filePath = dataPath || defaultDataPath();
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  const data = {
    version: 2,
    exportedAt: new Date().toISOString(),
    notes: notes,
    trash: trash
  };
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf-8');
  return true;
}

// 新增便签
function addNote(notes, data) {
  const note = createNote(data);
  notes.unshift(note);
  return { notes, note };
}

// 更新便签
function updateNote(notes, id, patch) {
  let updated = null;
  const newNotes = notes.map(n => {
    if (n.id === id) {
      updated = createNote({
        ...n,
        ...patch,
        id: n.id,
        createdAt: n.createdAt,
        updatedAt: Date.now()
      });
      return updated;
    }
    return n;
  });
  return { notes: newNotes, note: updated };
}

// 删除便签
function deleteNote(notes, id) {
  return notes.filter(n => n.id !== id);
}

// === 回收站操作 ===
// 回收站最长保留天数
const TRASH_MAX_DAYS = 30;

// 移入回收站：从 notes 移除并加入 trash（带 deletedAt 时间戳）
function moveToTrash(notes, trash, id) {
  const note = notes.find(n => n.id === id);
  if (!note) {
    return { notes, trash, note: null };
  }
  const trashed = createTrashNote({ ...note, deletedAt: Date.now() });
  const newNotes = notes.filter(n => n.id !== id);
  // 回收站按删除时间倒序（最新删除的在前面）
  const newTrash = [trashed, ...trash];
  return { notes: newNotes, trash: newTrash, note: trashed };
}

// 从回收站恢复：从 trash 移除并加回 notes（清除 deletedAt，取消置顶避免位置混乱）
function restoreNote(trash, notes, id) {
  const trashed = trash.find(n => n.id === id);
  if (!trashed) {
    return { notes, trash, note: null };
  }
  const restored = createNote({
    ...trashed,
    deletedAt: undefined,
    pinned: false, // 恢复后取消置顶，避免突兀置顶
    updatedAt: Date.now()
  });
  delete restored.deletedAt;
  const newTrash = trash.filter(n => n.id !== id);
  const newNotes = [restored, ...notes];
  return { notes: newNotes, trash: newTrash, note: restored };
}

// 从回收站彻底删除单条
function deleteFromTrash(trash, id) {
  return trash.filter(n => n.id !== id);
}

// 清空回收站
function emptyTrash(trash) {
  return [];
}

// 自动清理过期回收站（超过 TRASH_MAX_DAYS 天）
function autoCleanTrash(trash, now) {
  const nowTs = now || Date.now();
  const maxAgeMs = TRASH_MAX_DAYS * 24 * 60 * 60 * 1000;
  return trash.filter(n => {
    if (!n.deletedAt) return true;
    return (nowTs - n.deletedAt) < maxAgeMs;
  });
}

// 计算回收站便签的剩余保留天数
function getTrashDaysLeft(note, now) {
  const nowTs = now || Date.now();
  if (!note.deletedAt) return TRASH_MAX_DAYS;
  const maxAgeMs = TRASH_MAX_DAYS * 24 * 60 * 60 * 1000;
  const elapsed = nowTs - note.deletedAt;
  const left = Math.ceil((maxAgeMs - elapsed) / (24 * 60 * 60 * 1000));
  return left < 0 ? 0 : left;
}

// 切换置顶
function togglePin(notes, id) {
  let updated = null;
  const newNotes = notes.map(n => {
    if (n.id === id) {
      updated = { ...n, pinned: !n.pinned, updatedAt: Date.now() };
      return updated;
    }
    return n;
  });
  return { notes: newNotes, note: updated };
}

// 搜索便签（标题 + 内容，不区分大小写）
function searchNotes(notes, keyword) {
  if (!keyword || !keyword.trim()) {
    return notes;
  }
  const kw = keyword.trim().toLowerCase();
  return notes.filter(n =>
    (n.title || '').toLowerCase().includes(kw) ||
    (n.content || '').toLowerCase().includes(kw)
  );
}

// 按分类筛选
function filterByCategory(notes, category) {
  if (!category || category === '全部') {
    return notes;
  }
  return notes.filter(n => n.category === category);
}

// 排序：置顶优先 → 更新时间倒序
function sortNotes(notes) {
  return [...notes].sort((a, b) => {
    if (a.pinned !== b.pinned) {
      return b.pinned - a.pinned;
    }
    return b.updatedAt - a.updatedAt;
  });
}

// 统计信息
function getStats(notes) {
  const stats = {
    total: notes.length,
    pinned: notes.filter(n => n.pinned).length,
    byCategory: {},
    byColor: {},
    totalWords: 0,
    totalChars: 0
  };
  CATEGORIES.forEach(c => { stats.byCategory[c] = 0; });
  Object.keys(COLORS).forEach(c => { stats.byColor[c] = 0; });
  notes.forEach(n => {
    if (stats.byCategory[n.category] !== undefined) {
      stats.byCategory[n.category]++;
    }
    if (stats.byColor[n.color] !== undefined) {
      stats.byColor[n.color]++;
    }
    const text = (n.title + ' ' + n.content).trim();
    stats.totalChars += text.length;
    // 中英文混合字数统计：中文按字计，英文按词计
    const chinese = (text.match(/[\u4e00-\u9fa5]/g) || []).length;
    const english = (text.replace(/[\u4e00-\u9fa5]/g, ' ').match(/[a-zA-Z0-9]+/g) || []).length;
    stats.totalWords += chinese + english;
  });
  return stats;
}

// 导出为 JSON 字符串
function exportNotes(notes) {
  return JSON.stringify({
    version: 1,
    exportedAt: new Date().toISOString(),
    notes: notes
  }, null, 2);
}

// 从 JSON 字符串导入
function importNotes(jsonStr, existingNotes) {
  const data = JSON.parse(jsonStr);
  if (!Array.isArray(data.notes)) {
    throw new Error('无效的便签数据格式');
  }
  // 合并导入：保留已有便签，新增导入便签（重新生成 ID 避免冲突，保留原 updatedAt 时间戳）
  const imported = data.notes.map(n => createNote({
    ...n,
    id: undefined, // 重新生成 ID
    // 保留原 createdAt / updatedAt，避免导入后所有便签显示「刚刚」
    createdAt: n.createdAt,
    updatedAt: n.updatedAt
  }));
  return [...existingNotes, ...imported];
}

module.exports = {
  COLORS,
  CATEGORIES,
  TRASH_MAX_DAYS,
  generateId,
  createNote,
  createTrashNote,
  loadNotes,
  loadAll,
  saveNotes,
  saveAll,
  addNote,
  updateNote,
  deleteNote,
  moveToTrash,
  restoreNote,
  deleteFromTrash,
  emptyTrash,
  autoCleanTrash,
  getTrashDaysLeft,
  togglePin,
  searchNotes,
  filterByCategory,
  sortNotes,
  getStats,
  exportNotes,
  importNotes,
  defaultDataPath
};

File: sticky-notes-manager/portable/note_store.py
# -*- coding: utf-8 -*-
"""便签管家便携版 - 数据层
纯 Python 实现，零第三方依赖，与原 Electron 版数据格式兼容（version 2）。
"""

from __future__ import annotations

import json
import os
import time
import uuid
from typing import Any, Dict, List, Optional, Tuple

# === 常量 ===

# 颜色标签（与原版保持一致，苹果白风格配色）
COLORS: Dict[str, Dict[str, str]] = {
    "default": {"name": "默认", "hex": "#f5f5f7", "dot": "#8e8e93"},
    "blue":    {"name": "蓝色", "hex": "#e3f0ff", "dot": "#007aff"},
    "green":   {"name": "绿色", "hex": "#e8f8ec", "dot": "#34c759"},
    "yellow":  {"name": "黄色", "hex": "#fff9e0", "dot": "#ffcc00"},
    "orange":  {"name": "橙色", "hex": "#fff0e0", "dot": "#ff9500"},
    "pink":    {"name": "粉色", "hex": "#ffe8ef", "dot": "#ff2d55"},
    "purple":  {"name": "紫色", "hex": "#f3e8ff", "dot": "#af52de"},
}

CATEGORIES: List[str] = ["工作", "个人", "灵感", "待办", "其他"]

TRASH_MAX_DAYS = 30


def _now_ms() -> int:
    """当前毫秒时间戳"""
    return int(time.time() * 1000)


def generate_id() -> str:
    """生成唯一 ID"""
    return "n_" + uuid.uuid4().hex[:12]


def default_data_path() -> str:
    """默认数据文件路径（用户数据目录下，与原版路径不同，避免冲突）"""
    base = os.environ.get("APPDATA") or os.environ.get("HOME") or os.environ.get("USERPROFILE") or "."
    return os.path.join(base, "sticky-notes-portable", "notes.json")


def create_note(data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """创建便签对象（带默认值）"""
    data = data or {}
    now = _now_ms()
    color = data.get("color", "default")
    if color not in COLORS:
        color = "default"
    category = data.get("category", "其他")
    if category not in CATEGORIES:
        category = "其他"
    updated = data.get("updatedAt")
    if not (isinstance(updated, (int, float)) and updated > 0):
        updated = now
    return {
        "id": data.get("id") or generate_id(),
        "title": (data.get("title") or "").strip(),
        "content": (data.get("content") or "").strip(),
        "color": color,
        "category": category,
        "pinned": bool(data.get("pinned", False)),
        "createdAt": data.get("createdAt") or now,
        "updatedAt": updated,
    }


def create_trash_note(data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """创建回收站便签对象"""
    note = create_note(data)
    note["deletedAt"] = (data or {}).get("deletedAt") or _now_ms()
    return note


def load_all(data_path: Optional[str] = None) -> Dict[str, List[Dict[str, Any]]]:
    """读取便签 + 回收站"""
    path = data_path or default_data_path()
    if not os.path.exists(path):
        return {"notes": [], "trash": []}
    try:
        with open(path, "r", encoding="utf-8") as f:
            data = json.load(f)
        notes = [create_note(n) for n in data.get("notes", [])] if isinstance(data.get("notes"), list) else []
        trash = [create_trash_note(n) for n in data.get("trash", [])] if isinstance(data.get("trash"), list) else []
        return {"notes": notes, "trash": trash}
    except Exception:
        return {"notes": [], "trash": []}


def save_all(notes: List[Dict[str, Any]], trash: List[Dict[str, Any]], data_path: Optional[str] = None) -> bool:
    """保存全部便签 + 回收站"""
    path = data_path or default_data_path()
    os.makedirs(os.path.dirname(path), exist_ok=True)
    data = {
        "version": 2,
        "exportedAt": time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime()),
        "notes": notes,
        "trash": trash,
    }
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
    return True


def add_note(notes: List[Dict[str, Any]], data: Dict[str, Any]) -> Tuple[List[Dict[str, Any]], Dict[str, Any]]:
    """新增便签（插入到最前）"""
    note = create_note(data)
    return [note] + notes, note


def update_note(notes: List[Dict[str, Any]], note_id: str, patch: Dict[str, Any]) -> Tuple[List[Dict[str, Any]], Optional[Dict[str, Any]]]:
    """更新便签"""
    updated: Optional[Dict[str, Any]] = None
    new_notes: List[Dict[str, Any]] = []
    for n in notes:
        if n["id"] == note_id:
            merged = {**n, **patch, "id": n["id"], "createdAt": n["createdAt"], "updatedAt": _now_ms()}
            updated = create_note(merged)
            new_notes.append(updated)
        else:
            new_notes.append(n)
    return new_notes, updated


def toggle_pin(notes: List[Dict[str, Any]], note_id: str) -> Tuple[List[Dict[str, Any]], Optional[Dict[str, Any]]]:
    """切换置顶"""
    updated: Optional[Dict[str, Any]] = None
    new_notes: List[Dict[str, Any]] = []
    for n in notes:
        if n["id"] == note_id:
            updated = {**n, "pinned": not n["pinned"], "updatedAt": _now_ms()}
            new_notes.append(updated)
        else:
            new_notes.append(n)
    return new_notes, updated


def move_to_trash(notes: List[Dict[str, Any]], trash: List[Dict[str, Any]], note_id: str) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]], Optional[Dict[str, Any]]]:
    """移入回收站"""
    target = next((n for n in notes if n["id"] == note_id), None)
    if not target:
        return notes, trash, None
    trashed = create_trash_note({**target, "deletedAt": _now_ms()})
    return [n for n in notes if n["id"] != note_id], [trashed] + trash, trashed


def restore_note(trash: List[Dict[str, Any]], notes: List[Dict[str, Any]], note_id: str) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]], Optional[Dict[str, Any]]]:
    """从回收站恢复"""
    target = next((n for n in trash if n["id"] == note_id), None)
    if not target:
        return notes, trash, None
    restored = create_note({**target, "pinned": False, "updatedAt": _now_ms()})
    return [restored] + notes, [n for n in trash if n["id"] != note_id], restored


def delete_from_trash(trash: List[Dict[str, Any]], note_id: str) -> List[Dict[str, Any]]:
    """从回收站彻底删除"""
    return [n for n in trash if n["id"] != note_id]


def empty_trash(trash: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """清空回收站"""
    return []


def auto_clean_trash(trash: List[Dict[str, Any]], now_ms: Optional[int] = None) -> List[Dict[str, Any]]:
    """自动清理过期回收站"""
    now = now_ms or _now_ms()
    max_age_ms = TRASH_MAX_DAYS * 24 * 60 * 60 * 1000
    return [n for n in trash if not n.get("deletedAt") or (now - n["deletedAt"]) < max_age_ms]


def get_trash_days_left(note: Dict[str, Any], now_ms: Optional[int] = None) -> int:
    """计算回收站便签的剩余保留天数"""
    now = now_ms or _now_ms()
    if not note.get("deletedAt"):
        return TRASH_MAX_DAYS
    max_age_ms = TRASH_MAX_DAYS * 24 * 60 * 60 * 1000
    elapsed = now - note["deletedAt"]
    left = -(-(max_age_ms - elapsed) // (24 * 60 * 60 * 1000))  # 向上取整
    return max(0, int(left))


def search_notes(notes: List[Dict[str, Any]], keyword: str) -> List[Dict[str, Any]]:
    """搜索便签（标题 + 内容，不区分大小写"""
    if not keyword or not keyword.strip():
        return notes
    kw = keyword.strip().lower()
    return [n for n in notes if kw in (n.get("title") or "").lower() or kw in (n.get("content") or "").lower()]


def filter_by_category(notes: List[Dict[str, Any]], category: str) -> List[Dict[str, Any]]:
    """按分类筛选"""
    if not category or category == "全部":
        return notes
    return [n for n in notes if n.get("category") == category]


def sort_notes(notes: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """排序：置顶优先 → 更新时间倒序"""
    return sorted(notes, key=lambda n: (not n.get("pinned", False), -n.get("updatedAt", 0)))


def get_stats(notes: List[Dict[str, Any]]) -> Dict[str, Any]:
    """统计信息"""
    stats: Dict[str, Any] = {
        "total": len(notes),
        "pinned": sum(1 for n in notes if n.get("pinned")),
        "byCategory": {c: 0 for c in CATEGORIES},
        "byColor": {c: 0 for c in COLORS},
        "totalWords": 0,
        "totalChars": 0,
    }
    for n in notes:
        if n.get("category") in stats["byCategory"]:
            stats["byCategory"][n["category"]] += 1
        if n.get("color") in stats["byColor"]:
            stats["byColor"][n["color"]] += 1
        text = ((n.get("title") or "") + " " + (n.get("content") or "")).strip()
        stats["totalChars"] += len(text)
        # 简化版字数统计
        chinese = sum(1 for ch in text if "\u4e00" <= ch <= "\u9fa5")
        english = len([w for w in "".join(ch if not ("\u4e00" <= ch <= "\u9fa5") else " " for ch in text).split() if any(c.isalnum() for c in w)])
        stats["totalWords"] += chinese + english
    return stats


def export_notes(notes: List[Dict[str, Any]]) -> str:
    """导出为 JSON 字符串"""
    return json.dumps({
        "version": 1,
        "exportedAt": time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime()),
        "notes": notes,
    }, ensure_ascii=False, indent=2)


def import_notes(json_str: str, existing_notes: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """从 JSON 字符串导入（重新生成 ID，保留原时间戳）"""
    data = json.loads(json_str)
    if not isinstance(data.get("notes"), list):
        raise ValueError("无效的便签数据格式")
    imported: List[Dict[str, Any]] = []
    for n in data["notes"]:
        imported.append(create_note({
            **n,
            "id": None,  # 重新生成 ID
            "createdAt": n.get("createdAt"),
            "updatedAt": n.get("updatedAt"),
        }))
    return existing_notes + imported


def format_relative_time(ts: int, now_ms: Optional[int] = None) -> str:
    """格式化相对时间"""
    now = now_ms or _now_ms()
    diff = (now - ts) / 1000  # 秒
    if diff < 60:
        return "刚刚"
    if diff < 3600:
        return f"{int(diff // 60)} 分钟前"
    if diff < 86400:
        return f"{int(diff // 3600)} 小时前"
    if diff < 86400 * 30:
        return f"{int(diff // 86400)} 天前"
    return time.strftime("%Y-%m-%d", time.localtime(ts / 1000))
